File: dsa/Trapping Rainwater/trap.py
arr=[0,1,0,2,1,0,1,3,2,1,2,1]
n=len(arr)
prefix=[0]*n
suffix=[0]*n
prefix[0]=arr[0]
suffix[n-1]=arr[n-1]
for i in range(1,n):
    prefix[i]=max(prefix[i-1],arr[i])
for i in range(n-2,0,-1):
    suffix[i]=max(suffix[i+1],arr[i])
# print(prefix)
# print(suffix)
total=0
for i in range(n):
    leftmax=prefix[i]
    rightmax=suffix[i]
    if(arr[i]<leftmax and arr[i]<rightmax):
        total=total+min(leftmax,rightmax)-arr[i]
print(total)

File: dsa/Trapping Rainwater/Trap.java

public class Trap {
    public static void main(String[] args) {
        int[] arr={0,1,0,2,1,0,1,3,2,1,2,1};
        int n = arr.length;
        int[] prefix=new int[n];
        int[] suffix=new int[n];
        prefix[0]=arr[0];
        suffix[n-1]=arr[n-1];
        for(int i=1;i<n;i++){
            prefix[i]=Math.max(prefix[i-1],arr[i]);
        }
        for(int i=n-2;i>-1;i--){
            suffix[i]=Math.max(suffix[i+1],arr[i]);
        }
        int total=0;
        for(int i=0;i<n;i++){
            int leftmax=prefix[i];
            int rightmax=suffix[i];
            if(arr[i]<leftmax && arr[i]<rightmax){
                total+=Math.min(leftmax,rightmax)-arr[i];
            }
        }
        System.out.println(total);
    }
}
